cirq-core/cirq/experiments/readout_confusion_matrix_test.py

```diff
@@ -17,6 +17,35 @@
 import pytest
 
 from cirq.experiments.single_qubit_readout_calibration_test import NoisySingleQubitReadoutSampler
+from cirq.experiments.readout_confusion_matrix import TensoredConfusionMatrices
+
+
+def add_readout_error(
+    measurements: np.ndarray,
+    zero_errors: np.ndarray,
+    one_errors: np.ndarray,
+    rng: np.random.Generator = np.random.default_rng(),
+) -> np.ndarray:
+    """Add readout errors to measured (or simulated) bitstrings.
+
+    Args:
+        measurements: The bitstrings to which we will add readout errors. measurements[i,j] is the
+                      ith bitstring, qubit j.
+        zero_errors: zero_errors[i] is the probability of a 0->1 readout error on qubit i.
+        one_errors: one_errors[i] is the probability of a 1->0 readout error on qubit i.
+        rng: The pseudorandom number generator to use.
+
+    Returns:
+        New measurements but with readout errors added.
+    """
+    num_bitstrs, n = measurements.shape
+    assert len(zero_errors) == len(one_errors) == n
+    p1 = np.einsum("ij,j->ij", measurements, 1 - one_errors) + np.einsum(
```

Comment:
isn't this `p1 = measurements @ (1 - one_errors) + (1 - measurements) @ zero_errors`? if it's then the matrix product way is clearer to software engineers who are not necessarly familiar with Einstein summation

Reply:
No, it's different because in matrix multiplication the index `j` would be summed over, and here it is not.

Reply:
is there a way to do this without einsum? what is the function being calculated ? or is this some a reweighing of the matrix?

Reply:
It's computing the probability that the bit should be 1 after adding noise, for each of the measured bitstrings. The einsum is just saying
`p1[i,j] = measurements[i,j] * (1- one_errors)[j] + (1-measurements)[i,j] * zero_errors[j]`,
which you could also implement with a loop, but that would be less efficient.

Reply:
I prefer to keep the einsum if you don't mind.

Reply:
can't this be achieved by the broadcast behaviour of array multiplication? `p1 = measurements * (1 - one_errors) + (1 - measurements) * zero_errors` that is `*` instead of `@`. I tested this below and array multiplication is faster than einsum

you can make the choice here but if it's based on performance then array multiplication is faster. eitherway please add a comment explaining what the line does

```py3
In [1]: import numpy as np

In [2]: A = np.random.random((5000, 1 << 18))

In [3]: B = np.random.random(1 << 18)

In [4]: want = np.einsum('ij,j->ij', A, B)

In [5]: got = A * B

In [6]: np.allclose(want, got)
Out[6]: True

In [7]: %timeit A * B
2.61 s ± 61.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit np.einsum('ij,j->ij', A, B)
3.38 s ± 71 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Ok, you're right. I will change it, thanks.